README.md
# emacs

私の目的はemacsをich好みの設定にすること。<br> My purpose is to set emacs for ich

I'm [here](https://github.com/ich747/emacs.git)


- [x] 計画表
- [x] ベースinit.el
- [ ] package管理
  - [ ]leaf


eshell/history
exit
exit
exit
exit
exit
exit
exit
exit
exit
exit
exit
exit
ls
szsh
ls
ls
ls
ls


leafloader.el
(prog1 "prepare leaf"
  (prog1 "package"
    (custom-set-variables
     '(package-archives '(("org"   . "https://orgmode.org/elpa/")
                          ("melpa" . "https://melpa.org/packages/")
                          ("gnu"   . "https://elpa.gnu.org/packages/"))))
    (package-initialize))

  (prog1 "leaf"
    (unless (package-installed-p 'leaf)
      (unless (assoc 'leaf package-archive-contents)
        (package-refresh-contents))
      (condition-case err
          (package-install 'leaf)
        (error
         (package-refresh-contents)       ; renew local melpa cache if fail
         (package-install 'leaf))))

    (leaf leaf-keywords
      :ensure t
      :config (leaf-keywords-init)))

  (prog1 "optional packages for leaf-keywords"
    ;; optional packages if you want to use :hydra, :el-get,,,
    (leaf hydra :ensure t)
    (leaf el-get :ensure t
      :custom ((el-get-git-shallow-clone  . t)))))

myel/point-undo.el
;;; point-undo.el --- undo/redo position

;;  Copyright (C) 2006,2008 rubikitch <rubikitch atmark ruby-lang.org>
;;  Version: $Id: point-undo.el,v 1.6 2009/10/16 20:37:37 rubikitch Exp rubikitch $

;;  This program is free software; you can redistribute it and/or modify
;;  it under the terms of the GNU General Public License as published by
;;  the Free Software Foundation; either version 2 of the License, or
;;  (at your option) any later version.
;;    This program is distributed in the hope that it will be useful,
;;  but WITHOUT ANY WARRANTY; without even the implied warranty of
;;  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;;  GNU General Public License for more details.
;;    You should have received a copy of the GNU General Public License
;;  along with this program; if not, write to the Free Software
;;  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

;;; Commentary:

;; This package allows you to undo/redo point and window-start.
;; It is like w3m's UNDO/REDO commands.

;;; Commands:
;;
;; Below are complete command list:
;;
;;  `point-undo'
;;    Undo position.
;;  `point-redo'
;;    Redo position.
;;
;;; Customizable Options:
;;
;; Below are customizable option list:
;;

;;; Setup:

;; (require 'point-undo)
;; (define-key global-map [f5] 'point-undo)
;; (define-key global-map [f6] 'point-redo)

;;; Bug Report:
;;
;; If you have problem, send a bug report via M-x point-undo-send-bug-report.
;; The step is:
;;  0) Setup mail in Emacs, the easiest way is:
;;       (setq user-mail-address "your@mail.address")
;;       (setq user-full-name "Your Full Name")
;;       (setq smtpmail-smtp-server "your.smtp.server.jp")
;;       (setq mail-user-agent 'message-user-agent)
;;       (setq message-send-mail-function 'message-smtpmail-send-it)
;;  1) Be sure to use the LATEST version of point-undo.el.
;;  2) Enable debugger. M-x toggle-debug-on-error or (setq debug-on-error t)
;;  3) Use Lisp version instead of compiled one: (load "point-undo.el")
;;  4) Do it!
;;  5) If you got an error, please do not close *Backtrace* buffer.
;;  6) M-x point-undo-send-bug-report and M-x insert-buffer *Backtrace*
;;  7) Describe the bug using a precise recipe.
;;  8) Type C-c C-c to send.
;;  # If you are a Japanese, please write in Japanese:-)

;;; History:
;; 
;; $Log: point-undo.el,v $
;; Revision 1.6  2009/10/16 20:37:37  rubikitch
;; point-undo-list records position info only when point is moved.
;;
;; Revision 1.5  2008/12/27 15:21:03  rubikitch
;; *** empty log message ***
;;
;; Revision 1.4  2008/12/27 15:20:26  rubikitch
;; *** empty log message ***
;;
;; Revision 1.3  2008/12/27 15:19:38  rubikitch
;; refactoring
;;
;; Revision 1.2  2008/12/27 14:53:54  rubikitch
;; undo/redo not only point but also window-start.
;;

;; 2006/02/27: initial version

;;; Code:

(eval-when-compile (require 'cl))

(defvar point-undo-list nil)
(make-variable-buffer-local 'point-undo-list)

(defvar point-redo-list nil)
(make-variable-buffer-local 'point-redo-list)

(defun point-undo-pre-command-hook ()
  "Save positions before command."
  (unless (or (eq this-command 'point-undo)
              (eq this-command 'point-redo))
    
    (let ((cell (cons (point) (window-start))))
      (unless (equal cell (car point-undo-list))
       (setq point-undo-list (cons cell point-undo-list))))
    (setq point-redo-list nil)))
(add-hook 'pre-command-hook 'point-undo-pre-command-hook)

(defun point-undo-doit (list1 list2)
  ;; list1, list2 = {point-undo-list, point-redo-list}
  (destructuring-bind (pt . wst)
      (or (car (symbol-value list1)) '(nil)) ;nil-safe
    (when pt
      (set list1 (cdr (symbol-value list1)))
      (set list2 (cons (cons (point) (window-start)) (symbol-value list2)))
      (goto-char pt)
      (set-window-start (selected-window) wst))))

(defun point-undo ()
  "Undo position."
  (interactive)
  (point-undo-doit 'point-undo-list 'point-redo-list))

(defun point-redo ()
  "Redo position."
  (interactive)
  (when (or (eq last-command 'point-undo)
            (eq last-command 'point-redo))
    (point-undo-doit 'point-redo-list 'point-undo-list)))

;;;; Bug report
(defvar point-undo-maintainer-mail-address
  (concat "rubiki" "tch@ru" "by-lang.org"))
(defvar point-undo-bug-report-salutation
  "Describe bug below, using a precise recipe.

When I executed M-x ...

How to send a bug report:
  1) Be sure to use the LATEST version of point-undo.el.
  2) Enable debugger. M-x toggle-debug-on-error or (setq debug-on-error t)
  3) Use Lisp version instead of compiled one: (load \"point-undo.el\")
  4) If you got an error, please paste *Backtrace* buffer.
  5) Type C-c C-c to send.
# If you are a Japanese, please write in Japanese:-)")
(defun point-undo-send-bug-report ()
  (interactive)
  (reporter-submit-bug-report
   point-undo-maintainer-mail-address
   "point-undo.el"
   (apropos-internal "^point-undo-" 'boundp)
   nil nil
   point-undo-bug-report-salutation))

(provide 'point-undo)

;; How to save (DO NOT REMOVE!!)
;; (emacswiki-post "point-undo.el")
;;; point-undo.el ends here

init.el
;; -*- Mode: Emacs-Lisp ; Coding : utf-8 -*-


;;<先史>-----------------------------------------------*
;; <ディレクトリ構成>
;; <\ディレクトリ構成>

(eval-and-compile
  (customize-set-variable
   'package-archives '(("gnu"   . "https://elpa.gnu.org/packages/")
                       ("melpa" . "https://melpa.org/packages/")
                       ("org"   . "https://orgmode.org/elpa/")))
  (package-initialize)

  (unless (package-installed-p 'leaf)
    (package-refresh-contents)
    (package-install 'leaf))

  (leaf leaf-keywords
    :ensure t
    :init
    (leaf el-get :ensure t);:use github
    :config
    ;; initialize leaf-keywords.el
    (leaf-keywords-init)
    )
  (leaf bytecomp
    :custom
    ((byte-compile-warnings . '(not
                                obsolete
                                free-vars
                                unresolved
                                callargs
                                redefine
                                noruntime
                                cl-functions
                                interactive-only
                                make-local))
     (debug-on-error         . nil))
    :config
    (let ((win (get-buffer-window "*Compile-Log*")))
      (when win (delete-window win)))))



;;<\先史>----------------------------------------------*


;;<Pkg via leaf>---------------------------------------*

(leaf *USER
  :custom
  (user-full-name . "Hay Ich")
  (user-mail-address . "ich.hub.747@gmail.com")
  (user-login-name . "ich"))
(leaf cus-edit
  :doc "tools for customizing Emacs and Lisp packages"
  :tag "builtin" "faces" "help"
  :url "https://qiita.com/conao3/items/347d7e472afd0c58fbd7"
  :custom `((custom-file . ,(locate-user-emacs-file "custom.el"))))
(leaf *LEAF-CONVERT
  :doc "You can try to transform On *scratch*"
  :config
  (leaf use-package
    :tag "package"
    :ensure t
    :require t)
  (leaf leaf-convert
    :ensure t)
  (leaf leaf-tree
    :ensure t
    :custom ((imenu-list-size   . 30)
	     (imenu-list-position . 'left))))

(leaf FRAME
  :doc "フレームに関する設定"
  :config
  
  (leaf debug-on-error
    :doc "デバッグ表示"
    :custom (debug-on-error . t))
  
  (leaf start-up
    :doc "スタートアップの表示をしない"
    :custom
    (inhibit-startup-message . t)
    (inhibit-startup-screen  . t))
  
  (leaf title-bar
    :doc "タイトルバーにフルパス表示"
    :custom
    (frame-title-format . "%f"))
  
  (leaf tool-bar
    :doc "ツールバー要らない"
    :custom
    (tool-bar-mode . nil) :)

  (leaf tabbar
    :disabled t
    :tag "github"
    :ensure t)
  
  (leaf column-number
    :doc "行に関する変更"
    :config
    (leaf col-num
      :doc "行数をミニバッファ/左マージンに表示"
      :custom
      (column-number-mode . t)
      (global-linum-mode  . t)))
  
  (leaf days
    :doc "月日をミニバッファに表示"
    :custom
    (display-time-string-forms . '((format "%s/%s(%s)" month day dayname)))
    (display-time-mode         . t))
  
  (leaf y-or-n
    :doc "y or n"
    :config (defalias 'yes-or-no-p 'y-or-n-p)))

(leaf FRAME_DESIGN
  :tag "3rd_party"
  :doc "GNU Emacs のデザインを変更する"
  :config

  (leaf modus-themes
    :tag "gitlab" "github"
    :url
    "https://gitlab.com/protesilaos/modus-themes.git"
    "https://protesilaos.com/emacs/modus-themes#h:3ed03a48-20d8-4ce7-b214-0eb7e4c79abe"
    :emacs> 26.1
    :ensure t
    :bind ("<f3>" . modus-themes-toggle)
    )
)


(leaf WRITINGS
  :doc "書く設定"
  :config
  
  (leaf charactor
    :tag "builtin"
    :preface
    
    (leaf encode-decode
      :tag "encode" "decode" "Japanese"
      :custom ((set-language-environment . "Japanese")
	       (set-language-environment . 'utf-8)
	       (prefer-coding-system . 'utf-8)))
    :config
    
    (leaf font
      :doc "font指定"
      :config
      (leaf english
	:config
	(set-face-attribute 'default nil :family "Menlo" :height 128))
      (leaf japanese
	:config))
    
    (leaf char-scale
      :doc "文字幅"
      :custom
      (face-font-rescale-alist . '((".*Menlo.*" . 1.0) ("-cdac$" . 1.3)))))

  (leaf highlight
    :config

    (leaf hl-line
      :doc "ハイライト補完 現在の行"
      :custom-face
      ((hl-line quote ((t (:background "WhiteSmoke")))))
      :config (global-hl-line-mode t))
    
					;(leaf indent-guide
					;  :ensure t
					;  :custom
					;  (indent-guide-delay     . 0.0)
					;  (indent-guide-recursive . t))

    (leaf indent-tab
      :tag
      :doc "インデントを整えるコマンド"
      :config
      (leaf point-undo
	:tag "url" "myel"
	:url "http://www.emacswiki.org/cgi-bin/wiki/download/point-undo.el"
	:init
	(load "~/.emacs.d/myel/point-undo")
	:bind ([f7] . point-undo))
      
      (leaf all-indent
	:disabled t
	:req "point-undo"
	:after point-undo
	:preface
	(defun my:all-indent ()
	  (interactive)
	  (mark-whole-buffer)
	  (indent-region (region-beginning)(region-end))
	  (point-undo))
	:bind ("C-x C-j" . my:all-indent)))
    
    (leaf paren
      :tag "builtin"
      :doc "括弧の補完など"
      :custom
      (show-paren-delay   . 0.0)
      (show-paren-style   . 'expression)
      (show-paren-mode    . t)
      (electric-pair-mode . t)))
  )

(leaf FILE_SYSTEM
  :doc "ファイルの読み書き保存に関する設定"
  :config
  (leaf *saving-system
    :config
    
    (leaf lock-file
      :tag "builtin"
      :doc "ロックファイル作成をしない"
      :config (setq create-lockfiles nil))
    
    (leaf backup-file
      :tag "builtin"
      :doc "バックアップファイルの作成を/backupに変更"
      :custom
      `((auto-save-timeout              . 15)
	(auto-save-interval             . 60)
	(auto-save-file-name-transforms . '((".*" ,(locate-user-emacs-file "backup/") t)))
	(backup-directory-alist         . '((".*" . ,(locate-user-emacs-file "backup"))
					    (,tramp-file-name-regexp . nil)))
	(version-control                . t)
	(kept-new-versions              . 3)
	(kept-old-versions              . 1)
	(delete-old-versions            . t)
	(auto-save-list-file-prefix     . ,(locate-user-emacs-file "backup/.saves-"))))
    
    (leaf time-stamp
      :hook (before-save-hook . time-stamp)
      :custom
      ((time-stamp-active     . t)
       (time-stamp-line-limit . 10)
       (time-stamp-start      . "$Lastupdate: 2")
       (time-stamp-end        . "\\$")
       (time-stamp-format     . "%Y-%02m-%02d %02H:%02M:%02S")))
    
    (leaf revert-buffer
      :tag "builtin" "myconfig"
      :doc "現在のバッファの再読み込み"
      :preface
      (defun my:revert-buffer (&optional force-reverting)
	"revert-buffer-no-confirm-version"
	(interactive "P")
	(if (or force-reverting (not (buffer-modified-p)))
	    (revert-buffer :ignore-auto :noconfirm)
	  (error "The buffer has been modified")))
      :bind ("M-r" . my:revert-buffer))
    
    (leaf autorevert
      :custom
      ((auto-revert-interval . 0.1))
      :hook
      (emacs-startup-hook . global-auto-revert-mode)))
  )

(leaf *MODE
  :doc "特定のファイル形式に対応する"
  :config

  (leaf Markdown
    :disabled t
    :config
    (leaf markdown-preview-mode
      :el-get "ancane/markdown-preview-mode"
      :require t
      :commands markdown-preview-mode
      :custom
      (markdown-preview-stylesheets . (list "github.css")))
    
    (leaf  markdown-mode
      :ensure t
      :require t
      :custom (("README\\.md\\'" . gfm-mode)
               ("\\.md\\'" . markdown-mode)
               ("\\.markdown\\'" . markdown-mode))
      :setq (markdown-command . "multimarkdown")))

  (leaf Python)
  (leaf Rust)
  (leaf Csv_and_Json)
  )

(leaf KEYBIND
  :tag "builtin"
  :doc "便利キーバインド"
  :bind
  ("C-m"    . 'newline-and-indent)
  ("C-t"    . 'other-window)
  ("C-x k"  . 'kill-this-buffer)
  ("M-<up>" . 'switch-to-buffer)
  ("M-t"    . 'leaf-tree-mode)
  :custom
  (kill-whole-line . t)
  :config
  
  (leaf my:func-keybind
    :preface
    (defun my:enumerate-region (start end)
      "先頭が-から始まる行を箇条書きにする"
      (interactive "r")
      (save-excursion
	(let ((no 1));;no is Number of col
	  (goto-char start);;リージョンの先頭に移動
	  (while (re-search-forward "^-\\|^[0-9]+)" end t);;-から始まる行の検索
	    (replace-match (format "%d) " no))
	    (setq no (+ no 1)))))))
  :bind ("C-="    . my:enumerate-region))


(leaf flycheck
  :disabled t
  :doc "リアルタイムエラー表記。On-the-fly syntax checking"
  :req "dash-2.12.1" "pkg-info-0.4" "let-alist-1.0.4" "seq-1.11" "emacs-24.3"
  :tag "minor-mode" "tools" "languages" "convenience" "emacs>=24.3"
  :url "http://www.flycheck.org"
  :emacs>= 24.3
  :ensure t
  :bind (("M-n" . flycheck-next-error)
         ("M-p" . flycheck-previous-error))
  :global-minor-mode global-flycheck-mode)



(leaf ZONE
  :tag "builtin"
  :doc "zone拡張"
  :preface
  (defvar zone-programs [ zone-pgm-rotate-LR-variable ])
  (defun my:matrix ()
    "デフォルトを緑にしzoneする。終了後は戻してmessage"
    (interactive)
    (let ((default-color (face-foreground 'default)))
      (set-foreground-color "green")
      (zone)
      (set-foreground-color default-color)
      (message "They're watching you...")))
  :bind ("C-#" . my:matrix)
  :custom (zone-when-idle . 10))


(leaf *delete-file-if-no-contents
  :tag "拾い物"
  :url "https://uwabami.github.io/cc-env/Emacs.html"
  :doc "delete file if no contents"
  :preface
  (defun my:delf-no-contents ()
    (when (and (buffer-file-name (current-buffer))
	       (= (point-min) (point-max)))
      (delete-file (buffer-file-name (current-buffer)))))
  :hook (after-save-hook . my:delf-no-contents))



(leaf THIRD_PARTY
  :tag "3rd_party" "github"
  :config
  
  (leaf sky-color-clock
    :el-get "zk-phi/sky-color-clock"
    :require t
    :custom
    (sky-color-clock-format . "%H:%M")
    :config
    (sky-color-clock-initialize 33);FUK
    (push '(:eval (sky-color-clock)) (default-value 'mode-line-format))))



;;</Pkg via leaf>---------------------------------------*

;;exportするための宣言
(provide 'init)
